### PR TITLE
Fixes that the compiler crashes on invalid LINQ syntax

### LINKED ISSUE #10885: "csc.exe" exited with code -2146232797
**Version Used**:  1.2.0.60317

**Steps to Reproduce**:

```
using System.Collections.Generic;
using System.Linq;
namespace ConsoleApplication1
{
    class Program
    {
        public class AClass : List<string>
        {
        }
        static void Main(string[] args)
        {
            var b = (from a in AClass select a);
        }
    }
}
```

**Expected Behavior**:

Error message 

**Actual Behavior**:

Problem signature:
  Problem Event Name:	CLR20r3
  Problem Signature 01:	VBCSCompiler.exe
  Problem Signature 02:	1.2.0.60317
  Problem Signature 03:	d90b2aed
  Problem Signature 04:	Microsoft.CodeAnalysis.CSharp
  Problem Signature 05:	1.2.0.60317
  Problem Signature 06:	a2eab45f
  Problem Signature 07:	5e30
  Problem Signature 08:	3f2
  Problem Signature 09:	System.InvalidOperationException
  OS Version:	6.1.7601.2.1.0.256.4
  Locale ID:	1033
  Additional Information 1:	f3b8
  Additional Information 2:	f3b85aa92f370ccea20162cf3275b897
  Additional Information 3:	38ec
  Additional Information 4:	38ec5120b1c4325d71d35defcb892ddc

### LINKED ISSUE #11063: CSC crashes on invalid LINQ syntax
**Version Used**: 1.2.0.60317 (the version shipped with Visual Studio 2015 Update 2)

**Steps to Reproduce**:

1. Compile this code 
```csharp
using System;

namespace PureLogic
{
    public class Bag
    {
    }

    public static class SelectManyX
    {
        public static Bag Select(this Bag input, Func<int, int> func)
            => null;

        public static Bag A()
            => from r in Bag select 7;
    }
}
```

**Expected Behavior**:
Compilation error.

**Actual Behavior**:
Crash.

### PR DESCRIPTION
**Pull request by pdelvo**
Fixes #11063 and fixes #10885.

### CODE REVIEW

**Comment by mmitche** [comment]
testing ci, ignore:

@dotnet-bot test vsi p0

**Comment by mmitche** [comment]
@dotnet-bot test vsi p0

**Comment by mmitche** [comment]
@dotnet-bot test vsi p0

**Comment by mmitche** [comment]
@dotnet-bot test vsi p1

**Comment by mmitche** [comment]
@dotnet-bot test vsi

**Comment by mmitche** [comment]
@dotnet-bot test vsi p0 please

**Comment by mmitche** [comment]
@dotnet-bot test vsi p1

**Comment by mmitche** [comment]
@dotnet-bot test vsi p0 please

**Comment by mmitche** [comment]
@dotnet-bot test vsi p1 please

**Comment by mmitche** [comment]
@dotnet-bot test vsi please

**Comment by mmitche** [comment]
@dotnet-bot test vsi p0 please

**Comment by jmarolf** [comment]
@dotnet/roslyn-compiler

**Comment by gafter** [comment]
@khyperia I think these two bugs are assigned to you.

**Comment by khyperia** [comment]
Thanks for taking this on!

Unfortunately, I don't think the new behavior is correct. Two of the tests modified in this PR, `SelectFromType01` and `SelectFromType02` revert to the old, buggy behavior that the tests are designed to prevent. The URL linked in those tests (which I'm guessing you can't access?) specifies the "old behavior" is CS0119, and the expected behavior (and actual behavior after the fix) is no error. The commit fixing it is before first commit of this repo, so I can't link a specific change :(

I think the two bug reports are oriented around invoking an instance method (or extension method) as a static method with no receiver.

This is allowed:

```C#
class C
{
    public static object Select(System.Func<int, int> f) { return null; }
}

class M
{
    static void Main()
    {
        var dummy = from x in C select x;
        // basically calling C.Select(x => x), should work fine
    }
}
```

This is not, and causes a crash (as reported by the two issues):

```C#
class C
{
}

static class Ext
{
    public static object Select(this C c, System.Func<int, int> f) { return null; }
}

class M
{
    static void Main()
    {
        var dummy = from x in C select x;
        // Calling C.Select(x => x), doesn't work as it needs a value for the "c" parameter in the extension method
    }
}
```

File: src/Compilers/CSharp/Test/Semantic/Semantics/QueryTests.cs [hunk 1]
```diff
@@ -1421,7 +1421,10 @@ static void Main()
     static IEnumerable<T> Select<T>(Func<int, T> f) { return null; }
 }";
             var compilation = CreateCompilationWithMscorlibAndSystemCore(sourceCode);
-            compilation.VerifyDiagnostics();
+            compilation.VerifyDiagnostics(
+                // (8,27): error CS0119: 'C' is a type, which is not valid in the given context
+                //         var q = from x in C select x;
+                Diagnostic(ErrorCode.ERR_BadSKunknown, "C").WithArguments("C", "type").WithLocation(8, 27));
```
Comment (AlekseyTs) [inline src/Compilers/CSharp/Test/Semantic/Semantics/QueryTests.cs]: This looks like a breaking change from VS2013 compiler.

**Comment by gafter** [comment]
Because this PR breaks the semantics of Linq, we cannot accept it in this form.

**Pull request closed without merging**